### PR TITLE
 Unify coarsening properties for cell characteristics.

### PR DESCRIPTION
**Pull request by marcfehling**
Right now during coarsening, we determine the finite element on parent cells to be the least dominant one of their children in some cases, whilst picking the most dominant in others.

This PR unifies all implementations regarding cell properties to use the *least* dominant one (i.e. the one with the most degrees of freedom assigned). This may not be the optimal way to deal with this situation, but it certainly leads to less loss of accuracy. This affects picking the correct finite element for the transfer of solutions (`SolutionTransfer`, `p::d::SolutionTransfer`), the transfer of active finite element indices (`hp::DoFHandler`), as well as load balancing (`p::CellWeights`).

While unifying degrees of freedom, we stick to using the *most* dominant ones as described in the hp paper.

~Based on / Blocked by #8304.~
EDIT: Rebased and ready for review.

### CODE REVIEW

**Comment by marcfehling** [comment]
@bangerth @davydden -- We discussed this some time ago. I hope you agree to these changes.

File: source/distributed/cell_weights.cc [hunk 1]
```diff
@@ -143,19 +143,20 @@ namespace parallel
         case Triangulation<dim, spacedim>::CELL_COARSEN:
           {
             std::set<unsigned int> fe_indices_children;
-            for (unsigned int child_index = 0;
-                 child_index < GeometryInfo<dim>::max_children_per_cell;
```
Comment (davydden) [inline source/distributed/cell_weights.cc]: why did you need to change this? it's compile-time know, i would keep it.

Comment (marcfehling) [reply source/distributed/cell_weights.cc]: I changed that line in case we want to use anisotropic refinement.

I know that the `p::d::Triangulation` class does not support anisotropic refinement currently, but whenever we decide to use anisotropic refinement sometime in the future (when `p4est` supports it), we will look after those like needles in a haystack.

**Comment by marcfehling** [comment]
Fixed typo. Squashed commits. Testers passed.

File: source/distributed/cell_weights.cc (hunk 1 above)
Comment (bangerth) [reply source/distributed/cell_weights.cc]: Good find. I think you made the right call!
